### DISCUSSION TITLE
imbalanced_sampler

### DISCUSSION THREAD

**Discussion by zuluokonkwo** [pyg-team/pytorch_geometric#6366] [Q&A]
Hi,
I'm presently working on an imbalanced graph data. I tried using the torch geometric imbalanced sampler class to address this issue, but it always gives an error.

Scenario 1:  When I define a sampler with the whole dataset as in  "sampler=ImbalancedSampler(dataset)"  I get an out of index error when training with training data.
Scenario 2:  When I define a sampler with the training dataset  as in "sampler=ImbalancedSampler(train_dataset)" I get an assertion error. I understand what's going on in scenario 2  "assert len(dataset) == y.numel()". 

Is there a way around this?

**Comment by rusty1s** [comment]
Do you have a minimal example to reproduce? This would help to track down the issue.

**Reply by zuluokonkwo** [reply]
I've attached pictures of both scenarios.
For scenario one I get the index error during training. For scenario two, i get the error right after defining the sampler.

![scenario one](https://user-images.githubusercontent.com/106361071/211274534-24244d14-13e6-4dff-8a6c-0ee56312bec2.PNG)

![scenario two](https://user-images.githubusercontent.com/106361071/211274573-38d9e456-2586-4df1-b380-be12f896145c.PNG)

**Reply by rusty1s** [reply]
It is clear to me that scenario 1 fails - the sampler should operate on the same dataset. In scenario 2, I pushed a fix here: https://github.com/pyg-team/pytorch_geometric/pull/6374

**Reply by zuluokonkwo** [reply]
Thanks Matthias,
in Scenario 2, I can see you pushed a fix #6374 . However, I'm still getting the same error.

**Reply by rusty1s** [reply]
Did you install from master?

**Reply by zuluokonkwo** [reply]
No I didn’t.
Does the fix only work by reinstalling PyG?

**Reply by rusty1s** [reply]
Yes. Alternatively, you can patch it on your own by modifying the file in the
```python
import torch_geometric
print(torch_geometric.__file__)
```
path

**Reply by zuluokonkwo** [reply]
No worries man,
Thanks a lot.

**Reply by zuluokonkwo** [reply]
Still the same error. Sampling dosen't work on the sliced dataset.

**Reply by rusty1s** [reply]
Did you patch it on your own? What's the shape of `dataset.y` and `len(dataset)`?

**Reply by zuluokonkwo** [reply]
I patched it myself by manually editing the source code in torch_geometric.

`len(dataset)` = 18468
`len(train_dataset)` =12930
 `dataset.y` and `train_dataset.y` gives an attribute error.

This code `y=dataset.y.view(-1)` also gives an attribute error.

The sampler seems to have no effect on the sliced data, It keeps referencing the parent data.

**Reply by rusty1s** [reply]
Yeah, looks like you need to indeed install latest master, patching does not work. Sorry for the inconvenience!

**Reply by zuluokonkwo** [reply]
No worries.

**Reply by zuluokonkwo** [reply]
After installing from master the same error still pops up. Still can't perform sampling on the sliced dataset.

**Reply by rusty1s** [reply]
Are you sure the master is correctly picked up? You can try to re-install `torch-geometric` first (multiple times), and then install again. The following snippet works for me in master:
```python
dataset = TUDataset('/tmp/ENZYMES', name='ENZYMES')
dataset = dataset[:50]

sampler = ImbalancedSampler(dataset)
loader = DataLoader(dataset, batch_size=10, sampler=sampler)

for batch in loader:
    print(batch)

**Reply by zuluokonkwo** [reply]
Can you please change the name of the sliced data to something else and see if it works?

```
dataset = TUDataset('/tmp/ENZYMES', name='ENZYMES')
sliced_dataset = dataset[:50]

sampler = ImbalancedSampler(sliced_dataset )
loader = DataLoader(sliced_dataset , batch_size=10, sampler=sampler)

for batch in loader:
    print(batch
```

**Reply by rusty1s** [reply]
Still works.

**Reply by zuluokonkwo** [reply]
Its all good now,
Many thanks for your help Matthias.

**Comment by nasrinsaalehi** [comment]
Converting y Tensor to torch long will probably solve the assertion error.
```
data = Data(x=self.drug_encodings, y=y.to(torch.long), edge_index=self.edge_list)
# Sample the graph according to the ImbalancedSampler based on y distribution
sampler = ImbalancedSampler(dataset= data.y)
```
